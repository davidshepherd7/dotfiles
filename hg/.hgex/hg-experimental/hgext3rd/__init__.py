from __future__ import absolute_import
import pkgutil
# Indicate that hgext3rd is a namspace package, and other python path
# directories may still be searched for hgext3rd extensions.
__path__ = pkgutil.extend_path(__path__, __name__)

### IMPORTANT ###
# Do not add logic here that would diverge from mercurial's
# hgext3rd/__init__.py, the installed version of the file on debian systems is
# provided by mercurial itself; the packaging rules for remotefilelog explicitly
# ignore the file you're reading now.
