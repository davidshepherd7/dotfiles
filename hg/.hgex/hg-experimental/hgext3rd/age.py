# age.py
#
# Copyright 2017 Facebook, Inc.
#
# This software may be used and distributed according to the terms of the
# GNU General Public License version 2 or any later version.
"""
a revset predicate for filtering by changeset age.

Adds the `age()` revset predicate.

This revset predicate differs from the built-in `date` by providing a more
granular way of considering relative time rather than absolute time.

The built-in `date()` predicate does provide full day resolution, so
`age("<Xd")` is equivalent to `date("-X")`.
"""

from __future__ import absolute_import
import time
import re

from mercurial import (
    error,
    registrar,
    revsetlang,
)

revsetpredicate = registrar.revsetpredicate()

_rangeparser = re.compile(
        r'^([<>])(?:(\d+)d)?(?:(\d+)h)?(?:(\d+)m)?(?:(\d+)s?)?$')

@revsetpredicate('age(string)')
def age(repo, subset, x):
    """Changesets that are older or newer than a specific age.

    The age range can be specified in days, hours, minutes or seconds:

    - ``<30d``  : Newer than 30 days old
    - ``>4h30m``: Older than 4 hours 30 minutes old
    - ``<15s``  : Newer than 15 seconds old

    If no unit is specified, seconds are assumed.
    """
    agerange = revsetlang.getstring(x, 'age requires an age range')
    m = _rangeparser.match(agerange)
    if not m:
        raise error.ParseError('invalid age range for age predicate')
    dirn, days, hours, minutes, seconds = m.groups()
    cutoff = time.time()
    cutoff -= int(days or 0) * 60 * 60 * 24
    cutoff -= int(hours or 0) * 60 * 60
    cutoff -= int(minutes or 0) * 60
    cutoff -= int(seconds or 0)

    def newer(x):
        return repo[x].date()[0] > cutoff

    def older(x):
        return repo[x].date()[0] < cutoff

    if dirn == '<':
        return subset.filter(newer, condrepr=('<age %r>', agerange))
    else:
        return subset.filter(older, condrepr=('<age %r>', agerange))
