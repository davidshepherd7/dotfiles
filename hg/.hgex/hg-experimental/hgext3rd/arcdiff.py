# arcdiff.py - extension adding an option to the diff command to show changes
#              since the last arcanist diff
#
# Copyright 2016 Facebook, Inc.
#
# This software may be used and distributed according to the terms of the
# GNU General Public License version 2 or any later version.

import os

from mercurial import commands, error, extensions
from mercurial.i18n import _

from phabricator import (
    arcconfig,
    conduit,
    diffprops,
)

def extsetup(ui):
    entry = extensions.wrapcommand(commands.table, 'diff', _diff)
    options = entry[1]
    options.append(('', 'since-last-arc-diff', None,
        _('show changes since last `arc diff`')))

def _differentialhash(ui, repo, phabrev):
    client = conduit.Client()
    try:
        client.apply_arcconfig(arcconfig.load_for_path(repo.root))

        diffid = diffprops.getcurrentdiffidforrev(client, phabrev)
        if diffid is None:
            return None

        localcommits = diffprops.getlocalcommitfordiffid(client, diffid)
        return localcommits.get('commit', None) if localcommits else None
    except conduit.ClientError as e:
        ui.warn(_('Error calling conduit: %s\n') % str(e))
        return None
    except arcconfig.ArcConfigError as e:
        raise error.Abort(str(e))

def _diff(orig, ui, repo, *pats, **opts):
    if not opts.get('since_last_arc_diff'):
        return orig(ui, repo, *pats, **opts)

    ctx = repo['.']
    phabrev = diffprops.parserevfromcommitmsg(ctx.description())

    if phabrev is None:
        mess = _('local changeset is not associated with a differential '
                 'revision')
        raise error.Abort(mess)

    rev = _differentialhash(ui, repo, phabrev)
    if rev is None:
        mess = _('unable to determine previous changeset hash')
        raise error.Abort(mess)

    rev = str(rev)
    opts['rev'] = [rev]

    # if patterns aren't provided, restrict diff to files in both changesets
    # this prevents performing a diff on rebased changes
    if len(pats) == 0:
        prev = set(repo.unfiltered()[rev].files())
        curr = set(repo['.'].files())
        pats = tuple(os.path.join(repo.root, p) for p in prev | curr)

    return orig(ui, repo, *pats, **opts)
