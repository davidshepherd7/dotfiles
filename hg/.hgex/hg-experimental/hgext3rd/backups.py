# backups.py
#
# Copyright 2013 Facebook, Inc.
#
# This software may be used and distributed according to the terms of the
# GNU General Public License version 2 or any later version.
"""display recently made backups to recover stripped changesets

This extension allows the listing and restoring of strip backups.

By default, this will display a warning if you have obsolescence marker creation
enabled (since most hg installations use either strip or markers, but not both).
To disable that warning, set ``backups.warnobsolescence`` to False.
"""

from mercurial import cmdutil, commands, error, bundlerepo
from mercurial import hg, exchange, obsolete, registrar
from mercurial import bundle2, registrar
from mercurial import lock as lockmod
from mercurial import pycompat
from mercurial import scmutil
from mercurial import util
from hgext import pager
from mercurial.node import nullid, short
from mercurial.i18n import _
import os, glob, time

pager.attended.append('backups')

cmdtable = {}
command = registrar.command(cmdtable)
testedwith = 'ships-with-fb-hgext'
msgwithcreatermarkers = """Marker creation is enabled so no changeset should be
stripped unless you explicitly called hg strip. hg backups will show you the
stripped changesets. If you are trying to recover a changeset hidden from a
previous command, use hg journal to get its sha1 and you will be able to access
it directly without recovering a backup.\n\n"""
verbosetemplate = "{label('status.modified', node|short)} {desc|firstline}\n"

@command('^backups', [
    ('', 'recover', '',
     'brings the specified changeset back into the repository')
    ] + commands.logopts, _('hg backups [--recover HASH]'))
def backups(ui, repo, *pats, **opts):
    '''lists the changesets available in backup bundles

    Without any arguments, this command prints a list of the changesets in each
    backup bundle.

    --recover takes a changeset hash and unbundles the first bundle that
    contains that hash, which puts that changeset back in your repository.

    --verbose will print the entire commit message and the bundle path for that
    backup.
    '''
    supportsmarkers = obsolete.isenabled(repo, obsolete.createmarkersopt)
    if supportsmarkers and ui.configbool('backups', 'warnobsolescence', True):
        # Warn users of obsolescence markers that they probably don't want to
        # use backups but reflog instead
        ui.warn(msgwithcreatermarkers)
    backuppath = repo.vfs.join("strip-backup")
    backups = filter(os.path.isfile, glob.glob(backuppath + "/*.hg"))
    backups.sort(key=lambda x: os.path.getmtime(x), reverse=True)

    opts['bundle'] = ''
    opts['force'] = None

    if util.safehasattr(cmdutil, 'loglimit'):
        # legacy case
        loglimit = cmdutil.loglimit
        show_changeset = cmdutil.show_changeset
    else:
        # since core commit c8e2d6ed1f9e
        from mercurial import logcmdutil
        loglimit = logcmdutil.getlimit
        show_changeset = logcmdutil.changesetdisplayer

    def display(other, chlist, displayer):
        limit = loglimit(opts)
        if opts.get('newest_first'):
            chlist.reverse()
        count = 0
        for n in chlist:
            if limit is not None and count >= limit:
                break
            parents = [p for p in other.changelog.parents(n) if p != nullid]
            if opts.get('no_merges') and len(parents) == 2:
                continue
            count += 1
            displayer.show(other[n])

    recovernode = opts.get('recover')
    if recovernode:
        if scmutil.isrevsymbol(repo, recovernode):
            ui.warn(_("%s already exists in the repo\n") % recovernode)
            return
    else:
        msg = _('Recover changesets using: hg backups --recover '
                '<changeset hash>\n\nAvailable backup changesets:')
        ui.status(msg, label="status.removed")

    for backup in backups:
        # Much of this is copied from the hg incoming logic
        source = os.path.relpath(backup, pycompat.getcwd())
        source = ui.expandpath(source)
        source, branches = hg.parseurl(source, opts.get('branch'))
        try:
            other = hg.peer(repo, opts, source)
        except error.LookupError as ex:
            msg = _("\nwarning: unable to open bundle %s") % source
            hint = _("\n(missing parent rev %s)\n") % short(ex.name)
            ui.warn(msg)
            ui.warn(hint)
            continue
        revs, checkout = hg.addbranchrevs(repo, other, branches,
                                          opts.get('rev'))

        if revs:
            revs = [other.lookup(rev) for rev in revs]

        quiet = ui.quiet
        try:
            ui.quiet = True
            other, chlist, cleanupfn = bundlerepo.getremotechanges(ui, repo,
                                        other, revs, opts["bundle"],
                                        opts["force"])
        except error.LookupError:
            continue
        finally:
            ui.quiet = quiet

        try:
            if chlist:
                if recovernode:
                    tr = lock = None
                    try:
                        lock = repo.lock()
                        if scmutil.isrevsymbol(other, recovernode):
                            ui.status(_("Unbundling %s\n") % (recovernode))
                            f = hg.openpath(ui, source)
                            gen = exchange.readbundle(ui, f, source)
                            tr = repo.transaction("unbundle")
                            if not isinstance(gen, bundle2.unbundle20):
                                gen.apply(repo, 'unbundle', 'bundle:' + source)
                            if isinstance(gen, bundle2.unbundle20):
                                bundle2.applybundle(repo, gen, tr,
                                                    source='unbundle',
                                                    url='bundle:' + source)
                            tr.close()
                            break
                    finally:
                        lockmod.release(lock, tr)
                else:
                    backupdate = os.path.getmtime(source)
                    backupdate = time.strftime('%a %H:%M, %Y-%m-%d',
                                                time.localtime(backupdate))
                    ui.status("\n%s\n" % (backupdate.ljust(50)))
                    if not ui.verbose:
                        opts['template'] = verbosetemplate
                    else:
                        ui.status("%s%s\n" % ("bundle:".ljust(13), source))
                    displayer = show_changeset(ui, other, opts, False)
                    display(other, chlist, displayer)
                    displayer.close()
        finally:
            cleanupfn()
