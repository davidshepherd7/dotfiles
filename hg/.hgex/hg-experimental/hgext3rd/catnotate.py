from mercurial import util, cmdutil, registrar, scmutil, util
from mercurial.i18n import _
import os

cmdtable = {}
command = registrar.command(cmdtable)
testedwith = 'ships-with-fb-hgext'

@command('^catnotate', [
    ('r', 'rev', '', _('print the given revision'), _('REV')),
    ('a', 'text', None, _('treat all files as text')),
    ], _('[OPTION]... FILE...'),
)
def catnotate(ui, repo, file1, *args, **opts):
    """output the current or given revision of files annotated with filename
    and line number.

    Print the specified files as they were at the given revision. If
    no revision is given, the parent of the working directory is used.

    Binary files are skipped unless -a/--text option is provided.
    """
    ctx = scmutil.revsingle(repo, opts.get('rev'))
    matcher = scmutil.match(ctx, (file1,) + args, opts)
    prefix = ''

    err = 1
    # modified and stripped mercurial.cmdutil.cat follows
    def write(path):
        fp = cmdutil.makefileobj(repo, opts.get('output'), ctx.node(),
                         pathname=os.path.join(prefix, path))
        data = ctx[path].data()
        if not opts.get('text') and util.binary(data):
            fp.write("%s: binary file\n" % path)
            return

        for (num, line) in enumerate(data.split("\n"), start=1):
            line = line + "\n"
            fp.write("%s:%s: %s" % (path, num, line))
        fp.close()

    # Automation often uses hg cat on single files, so special case it
    # for performance to avoid the cost of parsing the manifest.
    if len(matcher.files()) == 1 and not matcher.anypats():
        file = matcher.files()[0]
        mfl = repo.manifestlog
        mfnode = ctx.manifestnode()
        if mfnode and mfl[mfnode].find(file)[0]:
            write(file)
            return 0

    for abs in ctx.walk(matcher):
        write(abs)
        err = 0

    return err
