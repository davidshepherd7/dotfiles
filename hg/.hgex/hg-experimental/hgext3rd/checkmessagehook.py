import string

from mercurial.i18n import _

def reposetup(ui, repo):
    ui.setconfig('hooks', 'pretxncommit.checkmessage', checkcommitmessage)

def checkcommitmessage(ui, repo, **kwargs):
    """
    Checks a single commit message for adherence to commit message rules.
    """
    hg_commit_message = repo['tip'].description()
    try:
        hg_commit_message.decode('utf8')
    except UnicodeDecodeError:
        ui.warn(_('commit message is not utf-8\n'))
        return True

    printable = set(string.printable)
    for c in hg_commit_message:
        if ord(c) < 128 and c not in printable:
            ui.warn(_('non-printable characters in commit message\n'))
            return True

    # False means success
    return False
