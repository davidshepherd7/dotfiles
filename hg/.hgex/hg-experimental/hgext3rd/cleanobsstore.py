# Copyright 2017 Facebook, Inc.
#
# This software may be used and distributed according to the terms of the
# GNU General Public License version 2 or any later version.
"""
extension for cleaning users' obsstore

this extension can be used for very narrow purpose. It deletes useless
obsmarkers from obsstore. 'Useless' is defined as marker whose username is in a
list of bad usernames (see config options below).
For example, if automation created lots of obsolete markers that were
accidentally pulled by many users then this extension can help. Probably the
better way to find useless markers is to check if precursor node is present in
the repo but it will be  slower.

Also note that this extension will run only once.

::

    [cleanobsstore]
    # list of usernames whose markers to delete from obsstore
    badusernames =
    # if obsstore size is bigger than obsstoresizelimit then we should try to
    # clean it
    obsstoresizelimit = 500000
"""

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

from mercurial import (
    obsutil,
    repair,
)
from mercurial.i18n import _

_cleanedobsstorefile = b'cleanedobsstore'

def reposetup(ui, repo):
    if repo.obsstore:
        indicestodelete = []
        if _needtoclean(ui, repo):
            repo._wlockfreeprefix.add(_cleanedobsstorefile)
            _write(ui, 'your obsstore is big, checking if we can clean it')
            badusernames = ui.configlist('cleanobsstore', 'badusernames')
            for index, data in enumerate(repo.obsstore._all):
                marker = obsutil.marker(repo, data)
                username = marker.metadata().get('user')
                if username:
                    if username in badusernames:
                        indicestodelete.append(index)
                # Intentionally marked as cleaned before the actual cleaning.
                # This is to avoid repeatedly iterating through the whole
                # obsstore for users who have many obsmarkers
                _markcleaned(repo)

        if indicestodelete:
            _write(ui, _('cleaning your obsstore to make hg faster; ' +
                         'it is a one-time operation, please wait...'))
            with repo.lock():
                repair.deleteobsmarkers(repo.obsstore, indicestodelete)

def _needtoclean(ui, repo):
    obsstoresizelimit = ui.configint('cleanobsstore', 'obsstoresizelimit',
                                     500000)
    return (repo.svfs.stat('obsstore').st_size >= obsstoresizelimit and
            not repo.vfs.exists(_cleanedobsstorefile))

def _markcleaned(repo):
    with repo.vfs(_cleanedobsstorefile, 'w') as f:
        f.write('1')  # any text will do

def _write(ui, msg):
    if ui.interactive() and not ui.plain('cleanobsstore'):
        ui.warn(_('%s\n') % msg)
