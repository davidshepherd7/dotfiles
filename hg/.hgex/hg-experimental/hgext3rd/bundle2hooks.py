# bundle2hooks.py - fix bundle2's support for hooks prior to lock acquisition.
#
# Copyright 2012 Facebook
#
# This software may be used and distributed according to the terms of the
# GNU General Public License version 2 or any later version.

'''Hooks arguments are typically stored on the transaction object.  However, we
may want to add hooks arguments without starting a transaction.  This allows us
to queue hook arguments on the bundle2 operation object.

'''

from __future__ import absolute_import

from mercurial import bundle2
from mercurial import error
from mercurial.i18n import _

from hgext3rd.extutil import replaceclass

def reposetup(ui, repo):
    @replaceclass(bundle2, 'bundleoperation')
    class bundleoperationhooked(bundle2.bundleoperation):
        def __init__(self, repo, transactiongetter, *args, **kwargs):
            def gettransaction():
                transaction = transactiongetter()

                if self.hookargs is not None:
                    # the ones added to the transaction supercede those added
                    # to the operation.
                    self.hookargs.update(transaction.hookargs)
                    transaction.hookargs = self.hookargs

                    # mark the hookargs as flushed.  further attempts to add to
                    # hookargs will result in an abort.
                    self.hookargs = None

                return transaction

            super(bundleoperationhooked, self).__init__(repo, gettransaction,
                                                        *args, **kwargs)

            self.hookargs = {}

        def addhookargs(self, hookargs):
            if self.hookargs is None:
                raise error.Abort(
                    _('attempted to add hooks to operation after transaction '
                      'started'))
            self.hookargs.update(hookargs)
