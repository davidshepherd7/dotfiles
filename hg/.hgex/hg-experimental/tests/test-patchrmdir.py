from __future__ import absolute_import, print_function

import os
import sys

# make it runnable directly without run-tests.py
sys.path[0:0] = [os.path.join(os.path.dirname(__file__), '..')]

if not sys.platform.startswith('linux'):
    sys.stderr.write('skipped: linux required\n')
    sys.exit(80)

from hgext3rd import patchrmdir
from mercurial import util

patchrmdir.uisetup(None)

testtmp = os.environ['TESTTMP']

d1 = os.path.join(testtmp, 'd1')
d2 = os.path.join(testtmp, 'd1', 'd2')

os.mkdir(d1)
os.mkdir(d2)

def tryfunc(func):
    try:
        func()
    except Exception as ex:
        # normalize the error message across platforms
        if 'Non-empty directory' not in str(ex):
            ex = '*'
        else:
            ex = 'Non-empty directory'
        print('  error: %s' % ex)
    else:
        print('  success')

print('rmdir d1 - should fail with ENOTEMPTY')
tryfunc(lambda: os.rmdir(d1))

print('rmdir d1/d2 - should succeed')
tryfunc(lambda: os.rmdir(d2))

open(d2, 'w').close()

print('rmdir d1 - should fail with ENOTEMPTY')
tryfunc(lambda: os.rmdir(d1))

os.unlink(d2)

print('rmdir d1 - should succeed')
tryfunc(lambda: os.rmdir(d1))

print('rmdir d1 - should fail with ENOENT')
tryfunc(lambda: os.rmdir(d1))

os.mkdir(d1)
os.mkdir(d2)

print('removedirs d2 (and d1) - should succeed')
tryfunc(lambda: util.removedirs(d2))

print('removedirs d1 - should fail with ENOENT')
tryfunc(lambda: util.removedirs(d1))
