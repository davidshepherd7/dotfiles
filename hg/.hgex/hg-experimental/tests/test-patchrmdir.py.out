rmdir d1 - should fail with ENOTEMPTY
  error: Non-empty directory
rmdir d1/d2 - should succeed
  success
rmdir d1 - should fail with ENOTEMPTY
  error: Non-empty directory
rmdir d1 - should succeed
  success
rmdir d1 - should fail with ENOENT
  error: *
removedirs d2 (and d1) - should succeed
  success
removedirs d1 - should fail with ENOENT
  error: *
